cpp/src/tsne/tsne_runner.cuh

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <raft/cudart_utils.h>
+#include <common/device_buffer.hpp>
+#include <cuml/common/logger.hpp>
+#include <cuml/manifold/common.hpp>
+#include <rmm/device_uvector.hpp>
+#include "distances.cuh"
+#include "exact_kernels.cuh"
+#include "utils.cuh"
+
+#include "barnes_hut.cuh"
+#include "exact_tsne.cuh"
+
+namespace ML {
+
+template <typename tsne_input, typename value_idx, typename value_t>
+class TSNE_runner {
+ public:
+  TSNE_runner(const raft::handle_t &handle_, tsne_input &input_,
+              knn_graph<value_idx, value_t> &k_graph_, const value_idx dim_,
+              const float theta_, const float epssq_, float perplexity_,
```

Comment:
This is good here- where all the hyperparams are consistently 32-bit. We should keep this consistent everywhere, such as the corresponding instance variables and the other functions in the algorithm.